### PR TITLE
CHE-8593: Organize imports and resolve conflicts

### PR DESCRIPTION
**Pull request by tolusha**
### Reference issue
https://github.com/eclipse/che/issues/8593

### CODE REVIEW

**Comment by tolusha** [comment]
This RP requires changes in eclipse/eclipse.jdt.ls
```
--- a/org.eclipse.jdt.ls.core/META-INF/MANIFEST.MF
+++ b/org.eclipse.jdt.ls.core/META-INF/MANIFEST.MF
@@ -37,6 +37,8 @@ Export-Package: org.eclipse.jdt.ls.core.internal;x-friends:="org.eclipse.jdt.ls.
  org.eclipse.jdt.ls.core.internal.commands;x-friends:="org.eclipse.jdt.ls.tests",
  org.eclipse.jdt.ls.core.internal.contentassist;x-friends:="org.eclipse.jdt.ls.tests",
  org.eclipse.jdt.ls.core.internal.corext.codemanipulation;x-friends:="org.eclipse.jdt.ls.tests",
+ org.eclipse.jdt.ls.core.internal.corrections,
+ org.eclipse.jdt.ls.core.internal.corrections.proposals,
  org.eclipse.jdt.ls.core.internal.handlers;x-friends:="org.eclipse.jdt.ls.tests",
  org.eclipse.jdt.ls.core.internal.javadoc;x-friends:="org.eclipse.jdt.ls.tests",
  org.eclipse.jdt.ls.core.internal.managers;x-friends:="org.eclipse.jdt.ls.tests"
```

**Comment by tolusha** [comment]
I will try.
The most concerning part is about changes in eclipse/eclipse.jdt.ls project. 
It is what we can do?

**Comment by tolusha** [comment]
@tsmaeder I am not sure if it is possible to compute the final text edits on the client side since the imports order matter.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/ImportConflicts.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.api.dto;
+
+import java.util.List;
+
+/** @author Anantolii Bazko */
+public class ImportConflicts {
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/ImportConflicts.java]: Do we need this class? Its' a pure container for a list.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/OrganizeImports.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  public OrganizeImports() {
+    this.choices = new ArrayList<>();
+  }
+
+  public OrganizeImports(String resourceUri) {
+    this.resourceUri = resourceUri;
+    this.choices = new ArrayList<>();
+  }
+
+  public OrganizeImports(String resourceUri, List<String> choices) {
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/OrganizeImports.java]: Please document what the values for "resourceUri" and "choices" can be. This is API, so documentation should be proper.
As I understand it, you have to put in fully qualified type names?

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/OrganizeImportsResult.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ */
+package org.eclipse.che.jdt.ls.extension.api.dto;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.eclipse.lsp4j.WorkspaceEdit;
+
+/** @author Anatolii Bazko */
+public class OrganizeImportsResult {
+  private List<ImportConflicts> importConflicts;
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/OrganizeImportsResult.java]: I think a List<List<String>> would be sufficient. They are not really conflicts, either, I "ambigousTypes" may be a more appropriate word.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/imports/OrganizeImportsCommand.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        JavaModelUtil.convertCommandParameter(arguments.get(0), OrganizeImports.class);
+
+    File file = new File(JDTUtils.toURI(organizeImports.getResourceUri()));
+    if (!file.exists()) {
+      return new OrganizeImportsResult();
+    }
+
+    if (file.isDirectory()) {
+      return doOrganizeImportsInDirectory(arguments, pm);
+    } else {
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/imports/OrganizeImportsCommand.java]: or do if (file.isDirectory()) { ...} else if (file.isFile()) {...} else { // error }

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/imports/OrganizeImportsCommand.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    TextChange textChange = proposal.getTextChange();
+    TextEdit edit = textChange.getEdit();
+    TextEditConverter converter = new TextEditConverter(cu, edit);
+    rootEdit.getChanges().put(JDTUtils.toURI(cu), converter.convert());
+  }
+
+  /** It is designed to preserve and resolve exposed conflicts. */
+  private static class ConflictResolverChooseImportQuery implements IChooseImportQuery {
+    private List<ImportConflicts> remainedConflicts;
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/imports/OrganizeImportsCommand.java]: "remainingConflicts"

**Comment by tolusha** [comment]
@tsmaeder It is ready to review again

**Pull request merged**